### PR TITLE
feat(cdk/tree): implement the various skeleton methods

### PR DESCRIPTION
**Pull request by BobobUnicorn**

### CODE REVIEW

**Comment by andrewseguin** [comment]
@BobobUnicorn Are these draft PRs still being worked on? Feel free to sync with me via Chat if you want guidance going forward on this tree redesign

**Comment by BobobUnicorn** [comment]
Yeah, I'd been out for a little while; I'll get these cleaned up for review.

File: src/cdk/tree/tree.ts [hunk 1]
```diff
@@ -414,6 +480,72 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
   _getChildrenAccessor() {
     return this.treeControl?.getChildren ?? this.childrenAccessor;
   }
+
+  private _getAllDescendants(): Observable<T[]> {
+    return merge(...(this._dataNodes?.map(dataNode => this._getDescendants(dataNode)) ?? []));
+  }
+
+  private _getDescendants(dataNode: T): Observable<T[]> {
```
Comment (andrewseguin) [inline src/cdk/tree/tree.ts:488]: It's not clear to me why these getters are returning observables. It seems to convert realtime results to an observable that never updates. Could these be synchronous functions?

File: src/cdk/tree/tree.ts [hunk 2]
```diff
@@ -414,6 +480,72 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
...
+        i++
+      ) {
+        results.push(this._dataNodes[i]);
+      }
+      return observableOf(results);
+    }
+    if (this.childrenAccessor) {
+      return this._getChildrenRecursively(dataNode).pipe(
+        reduce(
+          (memo: T[], next) => {
```
[the remark this answers is not in the document]
Comment (BobobUnicorn) [reply src/cdk/tree/tree.ts]: added more descriptive naming

File: src/cdk/tree/tree.ts [hunk 3]
```diff
@@ -414,6 +480,72 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
...
+            return memo;
+          },
+          [dataNode],
+        ),
+      );
+    }
+    throw getTreeControlMissingError();
+  }
+
+  private _getChildrenRecursively(dataNode: T): Observable<T[]> {
```
[the remark this answers is not in the document]
Comment (BobobUnicorn) [reply src/cdk/tree/tree.ts:536]: done

File: src/cdk/tree/tree.ts [hunk 4]
```diff
@@ -414,6 +480,72 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
...
+      switchMap(children => {
+        return concat(
+          observableOf(children),
+          ...children.map(child => this._getChildrenRecursively(child)),
+        );
+      }),
+    );
+  }
+
+  private _trackExpansionKey(dataNode: T): K {
```
[the remark this answers is not in the document]
Comment (BobobUnicorn) [reply src/cdk/tree/tree.ts]: done

File: src/cdk/tree/tree.ts [hunk 5]
```diff
@@ -414,6 +480,72 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
...
+    return this.expansionKey?.(dataNode) ?? (dataNode as unknown as K);
```
[the remark this answers is not in the document]
Comment (BobobUnicorn) [reply src/cdk/tree/tree.ts:560]: done

File: src/cdk/tree/tree.ts [hunk 6]
```diff
@@ -448,7 +580,14 @@ export class CdkTreeNode<T, K = T> implements FocusableOption, OnDestroy, OnInit
 
   @Input()
   get isExpanded(): boolean {
-    return !!this._tree.treeControl?.isExpanded(this._data);
+    return !!this._tree.isExpanded(this._data);
```
[the remark this answers is not in the document]
Comment (BobobUnicorn) [reply src/cdk/tree/tree.ts]: sure is :joy:

**Review by andrewseguin — APPROVED** [review]

**Comment by andrewseguin** [comment]
CI seems to be failing - want to try rebasing and trying it again?

**Comment by BobobUnicorn** [comment]
@andrewseguin I think the cdk-tree-revamp branch in this repo would also need to be rebased.

**Comment by andrewseguin** [comment]
@BobobUnicorn Did you need help rebasing the branch? Looks like this PR is in a distorted place with 2,000 files and 500 commits

**Comment by BobobUnicorn** [comment]
@andrewseguin I do not have permissions to update the branch in this repository

**Comment by andrewseguin** [comment]
Ah gotcha, would you like to open a PR from your local rebased branch that just updates this branch

**Comment by BobobUnicorn** [comment]
#25613 should do it

**Pull request merged**